zio-json/shared/src/main/scala/zio/json/decoder.scala

```diff
@@ -71,6 +71,113 @@ trait JsonDecoder[A] { self =>
+    delimiter: JsonStreamDelimiter = JsonStreamDelimiter.Array
+  ): ZTransducer[Blocking, Throwable, Char, A] =
+    ZTransducer {
+      for {
+        // format: off
+        runtime    <- ZIO.runtime[Any].toManaged_
+        inQueue    <- Queue.unbounded[Take[Nothing, Char]].toManaged_
+        outQueue   <- Queue.unbounded[Take[Throwable, A]].toManaged_
+        ended      <- Ref.makeManaged(false)
+        reader     <- ZManaged.fromAutoCloseable {
+                        ZIO.effectTotal {
+                          def readPull: Iterator[Chunk[Char]] =
+                            runtime.unsafeRun(inQueue.take)
+                              .fold(
+                                end   = Iterator.empty,
+                                error = _ => Iterator.empty, // impossible
+                                value = v => Iterator.single(v) ++ readPull
+                              )
+
+                          new zio.stream.internal.ZReader(Iterator.empty ++ readPull)
+                        }
+                      }
+        jsonReader <- ZManaged.fromAutoCloseable(ZIO.effectTotal(new WithRetractReader(reader)))
+        process    <- effectBlockingInterrupt {
+                        // Exceptions fall through and are pushed into the queue
+                        @tailrec def loop(atBeginning: Boolean): Unit = {
+                          val nextElem = try {
+                            if (atBeginning && delimiter == JsonStreamDelimiter.Array)  {
+                              Lexer.char(Nil, jsonReader, '[')
+                            } else {
+                              delimiter match {
+                                case JsonStreamDelimiter.Newline =>
+                                  jsonReader.readChar() match {
+                                    case '\r' =>
+                                      jsonReader.readChar() match {
+                                        case '\n' => ()
+                                        case _    => jsonReader.retract()
+                                      }
+                                    case '\n' => ()
+                                    case _    => jsonReader.retract()
+                                  }
+
+                               case JsonStreamDelimiter.Array =>
+                                  jsonReader.nextNonWhitespace() match {
+                                    case ',' | ']' => ()
+                                    case _         => jsonReader.retract()
+                                  }
+                              }
+                            }
+
+                            unsafeDecode(Nil, jsonReader)
+                          } catch {
+                            case t @ JsonDecoder.UnsafeJson(trace) =>
+                              throw new Exception(JsonError.render(trace))
+                          }
+
+                          runtime.unsafeRun(outQueue.offer(Take.single(nextElem)))
+
```

Comment:
Putting single characters on a queue is going to mean this is extremely slow.

Unrelated, can we leverage converting a stream to a reader?

Reply:
This is not a character, but the decoded element.

Reply:
> Unrelated, can we leverage converting a stream to a reader?

We use `zio.stream.internal.ZReader` here, as does `ZStream#toReader`

Reply:
Ah, I see, this makes sense!